### ISSUE TITLE
The warning in the latest SDK on implicit package refs for metapackages 

### ISSUE THREAD

**Issue by blackdwarf** [dotnet/cli#5346]
With the latest SDK, the package references for `Microsoft.NETCore.App` and `Netstandard.Library` are implied by the SDK. If a project that is being built contains an explicit reference to any of the above metapackages, we will issue a warning that the explicit reference is not needed. This could be problematic because users may not know what to do with it and also because it may fail the build based on the settings. 

Some of the solutions presented here are:

* Remove the warning from the SDK and continue as-is. 
* Remove the package reference in the migration. 

/cc @livarcocc @piotrpMSFT @onovotny @ardalis @julielerman

**Comment by piotrpMSFT** [comment]
I believe this is a dupe of https://github.com/dotnet/sdk/issues/666

@srivatsn and @dsplaisted should comment, but I'm assuming we're tracking the external SDK issue here unless I hear otherwise.

Also, the SDK bug number is ominous.

**Comment by piotrpMSFT** [comment]
We chatted about this offline. Since `<PackageReference>` is not the intended mechanism for describing the project's runtime version, we'll need to have migration handle converting the `dependency` element into the right csproj elements.

**Comment by livarcocc** [comment]
@piotrpMSFT What are the right csproj elements (I hope element)?

**Comment by dsplaisted** [comment]
Here's my recollection of the migrate behavior we discussed today:

- If the target framework from project.json is .NET Standard, than don't include the NETStandard.Library package reference in the migrated .csproj.
- If the target framework from project.json is netcoreapp, then don't include the Microsoft.NETCore.App package reference in the migrated .csproj.
  - If the package version of the Microsoft.NETCore.App dependency in project.json isn't the same as the target framework version, then add a `RuntimeFrameworkVersion` property to the migrated .csproj with the value set to the Microsoft.NETCore.App package version from project.json.

Basically, apps can generally just specify the target framework and everything should work.  If apps want to specify a specific patch version of .NET Core, they can do so with the `RuntimeFrameworkVersion` property.  In project.json, the patch version of the runtime to use would be specified by the version of the Microsoft.NETCore.App package reference, so migration can take the package version and put it in that property.

See here for more details: https://github.com/dotnet/sdk/pull/633#issuecomment-272515440

**Comment by livarcocc** [comment]
@dsplaisted Why not include the netstandard.library package when migrating? What if the user is targeting say, 1.6.2 or some other version higher up than what the SDK has? Migration will put a package reference for it in the csproj. Will the user still get the warning then?

Other than that, we can make this change in to migration.

**Comment by dsplaisted** [comment]
@livarcocc My understanding is that there are no plans to ship a 1.6.2 version of the package NETStandard.Library, and that the prerelease versions of the package that exist now are just artifacts of the development process as we are working on .NET Standard 2.0.

So it doesn't seem important for dotnet migrate to support migrating projects targeting a version of NETStandard.Library higher than 1.6.2.  If you do want to support it, you can set the `NetStandardImplicitPackageVersion` property to the version of the package reference if it is higher than 1.6.1, and not include the PackageReference in the project.  If the version of the package reference is 1.6.1 or less, you would not include the PackageReference or set the `NetStandardImplicitPackageVersion` property.

**Comment by livarcocc** [comment]
@dsplaisted I can do that. But, even though I understand that there is no plan to ship 1.6.2, there are projects out there today (in project.json) that do target it. i.e. The vast majority of ASP.NET projects and we can not simply assume to break them.

Though, I am glad we have an alternative in the NetStandardImplicitPackageVersion. I will use that.

**Comment by BladeWise** [comment]
I suppose that the existence and purpose of `NetStandardImplicitPackageVersion` property should be documented [here](https://aka.ms/sdkimplicitrefs).
I just faced this issue with vs2017 RC3 targeting some Mircosoft pre-release packages and had to dig quite a bit to understand how to resolve the issue in the proper way.

As far as I understand at the moment there are 2 ways:
1. Set `NetStandardImplicitPackageVersion` to the expected version (I noticed it works with wildcard, I was able to set version as `1.6.2-*`).
2. Add the `PackageReference` directly (which seems to be is discouraged), optionally setting `DisableImplicitFrameworkReferences` to `True`.
It seems the proper way, for the time being, is the first, and should only hit projects referencing pre-release packages that need a specific NETStandard.Library verison..

Notice that I found an inconsistent behaviour regarding the warning
> A PackageReference for 'NETStandard.Library' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs

It appeared consistently in a project that was previously created with vs2017 RC2; it never appeared with a project created directly with RC3 and simply accepted an explicit reference to NETStandard.Library, without having to set `DisableImplicitFrameworkReferences`.

**Comment by livarcocc** [comment]
@BladeWise please, file an issue with your comments at the dotnet/sdk repo.

**Comment by mbrookson** [comment]
I'm seeing this 'issue' about package references with the dotnet SDK 2.1.4.

Is there something I need to do to remove these references as they are supposedly inferred? Or is there an update coming which stops this warning appearing?

**Comment by livarcocc** [comment]
Remove explicitly package references to Microsoft.NetCore.App or NetStandard.Library from your project.

**Comment by haf** [comment]
@livarcocc  There are no references to those (or I'd remove them just from reading the error message):

```
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AssemblyName>Logary.CSharp.Tests</AssemblyName>
    <Version>5.0.0-alpha</Version>
    <TargetFrameworks>net461;netcoreapp2.0</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup>
    <None Include="paket.references" />
    <ProjectReference Include="..\Logary.CSharp\Logary.CSharp.csproj" />
    <ProjectReference Include="..\Logary\Logary.fsproj" />
  </ItemGroup>
  <Import Project="..\..\.paket\Paket.Restore.targets" />
</Project>
```

**Comment by livarcocc** [comment]
Have you checked that Paket.Restore.targets is not adding them?

If not, you can run the following commands and attach the dg and binlog files so that we can investigate more:

```
dotnet msbuild /t:GenerateRestoreGraphFile /p:RestoreGraphOutputPath=a.dg
dotnet msbuild /t:GenerateRestoreGraphFile /bl
```

**Comment by haf** [comment]
@livarcocc No, I haven't. Let me see.

```
dotnet msbuild /t:GenerateRestoreGraphFile /p:RestoreGraphOutputPath=a.dg                                                  h@catamorphism
Microsoft (R) Build Engine version 15.6.82.30579 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

/usr/local/share/dotnet/sdk/2.1.101/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.DefaultItems.targets(198,5): warning : A PackageReference for 'Microsoft.NETCore.App' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [src/Logary.CSharp.Tests/Logary.CSharp.Tests.csproj]
  Writing src/Logary.CSharp.Tests/a.dg


» dotnet msbuild /t:GenerateRestoreGraphFile /bl                                                                            h@catamorphism
Microsoft (R) Build Engine version 15.6.82.30579 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

/usr/local/share/dotnet/sdk/2.1.101/MSBuild.dll /bl /m /t:GenerateRestoreGraphFile /v:m ./Logary.CSharp.Tests.csproj
/usr/local/share/dotnet/sdk/2.1.101/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.DefaultItems.targets(198,5): warning : A PackageReference for 'Microsoft.NETCore.App' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [src/Logary.CSharp.Tests/Logary.CSharp.Tests.csproj]
/usr/local/share/dotnet/sdk/2.1.101/NuGet.targets(122,5): error : Missing RestoreGraphOutputPath property! [src/Logary.CSharp.Tests/Logary.CSharp.Tests.csproj]
------------------------------
```

[a.zip](https://github.com/dotnet/cli/files/1828676/a.zip)

**Comment by haf** [comment]
@TheRealPiotrP Wouldn't it be a good idea to reopen this issue now that you have a repro with output/debug symbols?

**Comment by livarcocc** [comment]
@TheAngryByrd that happens to you because when packaging for netcoreapp2.0 you are setting `OutputType` to `exe`.

In that case, we don't mark the implicit dependency as `PrivateAssets=All` and the package shows up in your nuspec. In the case of apps, it is supposed to not be privateassets.

**Comment by TheAngryByrd** [comment]
@livarcocc good to know, thanks!

Ok so then

```
/usr/share/dotnet/sdk/2.1.101/Sdks/Microsoft.NET.Sdk/build/Microsoft.NET.Sdk.DefaultItems.targets(198,5): warning : A PackageReference for 'Microsoft.NETCore.App' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/tmp/b1dec36ea3504378addc8ada0006b363/MyCoolLib/tests/MyCoolLib.Tests/MyCoolLib.Tests.fsproj]
```

 https://travis-ci.org/TheAngryByrd/MiniScaffold/jobs/361733228#L2425 is a paket issue?

**Comment by livarcocc** [comment]
Not sure. Can't really tell just from the build. I think there is a higher chance that it is something to do with your build itself.

**Comment by TheAngryByrd** [comment]
Yeah I think my issue is paket specifically.  A project without it doesn't give me that warning.  I'll open an issue over there.
